lib/blocktree/blocktree.go

```diff
@@ -297,29 +301,53 @@ func (bt *BlockTree) Leaves() []Hash {
 	return la
 }
 
-// HighestCommonAncestor returns the highest block that is a Ancestor to both a and b
-func (bt *BlockTree) HighestCommonAncestor(a, b Hash) (Hash, error) {
+// LowestCommonAncestor returns the lowest common ancestor between two blocks in the tree.
+func (bt *BlockTree) LowestCommonAncestor(a, b Hash) (Hash, error) {
 	bt.RLock()
 	defer bt.RUnlock()
 
-	an := bt.getNode(a)
-	if an == nil {
+	aNode := bt.getNode(a)
+	if aNode == nil {
 		return common.Hash{}, ErrNodeNotFound
 	}
 
-	bn := bt.getNode(b)
-	if bn == nil {
+	bNode := bt.getNode(b)
+	if bNode == nil {
 		return common.Hash{}, ErrNodeNotFound
 	}
-
-	ancestor := an.highestCommonAncestor(bn)
-	if ancestor == nil {
-		// this case shouldn't happen - any two nodes in the blocktree must
-		// have a common ancestor, the lowest of which is the root node
-		return common.Hash{}, fmt.Errorf("%w: %s and %s", ErrNoCommonAncestor, a, b)
+	return lowestCommonAncestor(aNode, bNode)
+}
+func lowestCommonAncestor(aNode, bNode *node) (Hash, error) {
+	var higherNode *node
+	var lowerNode *node
+	if aNode.number > bNode.number {
+		higherNode = aNode
+		lowerNode = bNode
+	} else {
+		higherNode = bNode
+		lowerNode = aNode
+	}
+
+	higherNum := higherNode.number
+	lowerNum := lowerNode.number
+	diff := higherNum - lowerNum
+	for diff > 0 {
+		if higherNode.parent == nil {
+			return common.Hash{}, fmt.Errorf("%w: for block number %v", errorAncestorOutOfBoundsCheck, higherNum)
```

Comment:
Should we panic here instead of returning an error? Same for below check?
That way we can simplify the signature to not return an error

Unless this can possibly happen

Reply:
Duplicate comment sorry